client/landing/stepper/declarative-flow/internals/types.ts

```diff
@@ -245,25 +256,73 @@ export type FlowV2 = {
 
 export type Flow = FlowV1 | FlowV2;
 
-export type StepProps = {
-	navigation: NavigationControls;
-	stepName: string;
-	flow: string;
-	/**
-	 * If this is a step of a flow that extends another, pass the variantSlug of the variant flow, it can come handy.
-	 */
-	variantSlug?: string;
-	data?: StepperInternal.State[ 'stepData' ];
-	children?: React.ReactNode;
-	/**
-	 * These two prop are used internally by the Stepper to redirect the user from the user step.
-	 */
-	redirectTo?: string;
-	signupUrl?: string;
+/**
+ * This is a helper type to intersect A and B only if B is not never. Intersecting with never results in never which is not what we want.
+ */
+type ConditionalIntersection< TA, TB > = [ TB ] extends [ never ] ? TA : TA & TB;
+
+/**
+ * This is the type of the props passed to the step.
+ * @template StepDataShape - The types of the step submitted data.
+ * @example
+ * const step = ( props: StepProps< { submits: { siteSlug: string } } > ) => {
+ *   return <div>{ props.navigation.submit( { siteSlug: 'example.wordpress.com' } ) }</div>;
+ * };
+ */
+export type StepProps< StepDataShape extends StepPropTypes | never = never > =
+	ConditionalIntersection<
+		{
+			navigation: NavigationControls< StepDataShape[ 'submits' ] >;
+			stepName: string;
+			flow: string;
+			/**
+			 * If this is a step of a flow that extends another, pass the variantSlug of the variant flow, it can come handy.
+			 */
+			variantSlug?: string;
+			data?: StepperInternal.State[ 'stepData' ];
+			children?: React.ReactNode;
+			/**
+			 * These two prop are used internally by the Stepper to redirect the user from the user step.
+			 */
+			redirectTo?: string;
+			signupUrl?: string;
+		},
+		StepDataShape[ 'accepts' ]
+	>;
+
+/**
+ * This is the type of the step submitted and accepted props.
+ * @example
+ * const step: Step< { submits: { newUserName: string }, accepts: { userName: string } } > ) => {
+ *   return (
+ *     <div>
+ *       <h1>Hi {userName}!</h1>
+ *       <input onChange={ value => props.navigation.submit( { newUserName: value } ) } />
+ *     </div>
+ */
+type StepPropTypes = {
+	readonly submits?: Record< string, unknown >;
+	readonly accepts?: Record< string, unknown >;
 };
 
-export type Step = React.FC< StepProps >;
+/**
+ * This is the type of the step component.
+ * @template ConfiguredStepPropTypes - The types of the step submitted and accepted props.
+ * @example
+ * const step: Step< { submits: { newUserName: string }, accepts: { userName: string } } > ) => {
+ *   return (
+ *     <div>
+ *       <h1>Hi {userName}!</h1>
+ *       <input onChange={ value => props.navigation.submit( { newUserName: value } ) } />
+ *     </div>
+ *   );
+ * };
+ */
+export type Step< ConfiguredStepPropTypes extends StepPropTypes = never > = React.FC<
+	StepProps< ConfiguredStepPropTypes >
+>;
 
+// TODO: get rid of these. Every type should be specific.
```

Comment:
This feels like it should be another explicit task.

Reply:
https://github.com/Automattic/wp-calypso/issues/101498